File: Tig.java
import java.io.*;
import java.nio.charset.MalformedInputException;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

public class Tig {
    private static final String REPO_DIR = "repo_java";
    private static final String STAGED_FILE = "staged.csv";
    private static final String COMMIT_FILE = "commits.csv";
    private static final String TIGIGNORE_FILE = ".tigignore";
    private static final int HASH_LEN = 16;

    public static void main(String[] args) throws IOException {
        if (args.length < 1) {
            System.err.println("Usage: java Tig <command> [args]");
            return;
        }

        String command = args[0];
        switch (command) {
            case "init":
                if (args.length < 2) {
                    System.err.println("Usage: java Tig init <directory>");
                    return;
                }
                init(args[1]);
                break;

            case "add":
                if (args.length < 2) {
                    System.err.println("Usage: java Tig add <file>");
                    return;
                }
                add(args[1]);
                break;

            case "commit":
                if (args.length < 2) {
                    System.err.println("Usage: java Tig commit <message>");
                    return;
                }
                commit(args[1]);
                break;

            case "log":
                if (args.length < 1) {
                    System.err.println("Usage: java Tig log [-num]");
                    return;
                } else if (args.length < 2) {
                    log();  
                } else {
                    log(args[1]); 
                }
                break;
            
      

            case "status":
                status(System.getProperty("user.dir"));
                break;

            case "checkout":
                if (args.length < 2) {
                    System.err.println("Usage: java Tig checkout <commit_id>");
                    return;
                }
                checkout(args[1]);
                break;

            case "diff":
                if (args.length < 2) {
                    System.err.println("Usage: java Tig diff <filename>");
                    return;
                }
                diff(args[1]);
                break;

            default:
                System.err.println("Unknown command: " + command);
        }
    }

    private static void init(String repoPathStr) throws IOException {
        Path repoPath = Paths.get(repoPathStr);
        Path tigDir = repoPath.resolve(".tig");
        if (!Files.exists(repoPath)) {
            Files.createDirectories(repoPath);
            System.out.println("Created repository directory: " + repoPath.toAbsolutePath());
        }
    
        if (!Files.exists(tigDir)) {
            Files.createDirectory(tigDir);
            System.out.println("Initialized empty repository in: " + tigDir.toAbsolutePath());
        }
    
        Path commitsFile = tigDir.resolve(COMMIT_FILE);
        if (!Files.exists(commitsFile)) {
            Files.createFile(commitsFile);
            System.out.println("Created commits file: " + commitsFile.toAbsolutePath());
        }
    }

    private static void add(String fileName) throws IOException {
        Path repoPath = Paths.get(System.getProperty("user.dir"));
        Path tigDir = repoPath.resolve(".tig");
        Path stagedFile = tigDir.resolve(STAGED_FILE);
        Path tigignorePath = tigDir.resolve(TIGIGNORE_FILE);

        Set<String> ignoredFiles = loadTigignore(tigignorePath);

        if (ignoredFiles.contains(fileName)) {
            System.err.println("Error: File is ignored: " + fileName);
            return;
        }

        Path filePath = repoPath.resolve(fileName);
        if (!Files.exists(filePath)) {
            System.err.println("Error: File does not exist: " + fileName);
            return;
        }

        String fileHash = calculateHash(filePath);
        Map<String, String> stagedFiles = loadCsvToMap(stagedFile.toString());
        stagedFiles.put(fileName, fileHash);
        saveMapToCsv(stagedFile.toString(), stagedFiles);

        System.out.println("File staged: " + fileName);
    }

    private static void commit(String message) throws IOException {
        Path repoPath = Paths.get(System.getProperty("user.dir"));
        Path tigDir = repoPath.resolve(".tig");
        Path stagedFile = tigDir.resolve(STAGED_FILE);
        Path commitsFile = tigDir.resolve(COMMIT_FILE);
    
        Map<String, String> stagedFiles = loadCsvToMap(stagedFile.toString());
        if (stagedFiles.isEmpty()) {
            System.err.println("No files staged for commit.");
            return;
        }
    
        String commitId = UUID.randomUUID().toString().substring(0, 16);
        Path commitDir = tigDir.resolve(commitId);
        Files.createDirectory(commitDir);
    
        for (Map.Entry<String, String> entry : stagedFiles.entrySet()) {
            Path sourcePath = repoPath.resolve(entry.getKey());
            Path destPath = commitDir.resolve(entry.getValue() + ".bck");
            Files.copy(sourcePath, destPath);
        }
    
        try (BufferedWriter writer = Files.newBufferedWriter(commitsFile, StandardOpenOption.CREATE, StandardOpenOption.APPEND)) {
            String fileHashes = String.join(" ", stagedFiles.values());
            writer.write(String.join(",", commitId, Instant.now().toString(), message, String.join(" ", stagedFiles.keySet()), fileHashes));
            writer.newLine();
        }
    
        Files.delete(stagedFile);
        System.out.println("Committed changes with ID: " + commitId);
    }

    private static void status(String repoPathStr) throws IOException {
        Path repoPath = Paths.get(repoPathStr);
        Path tigDir = repoPath.resolve(".tig");
        Path stagedFile = tigDir.resolve(STAGED_FILE);
        Path tigignorePath = tigDir.resolve(TIGIGNORE_FILE);
        Path commitsFile = tigDir.resolve(COMMIT_FILE);
    
        Set<String> ignoredFiles = loadTigignore(tigignorePath);
        Map<String, String> stagedFiles = loadCsvToMap(stagedFile.toString());
        List<Path> workingFiles = Files.walk(repoPath)
                .filter(Files::isRegularFile)
                .filter(file -> !file.startsWith(tigDir))
                .collect(Collectors.toList());
    
        List<String> allCommittedFiles = new ArrayList<>();
        List<String> allCommittedHashes = new ArrayList<>();

        // Building commit state
        try (BufferedReader reader = Files.newBufferedReader(commitsFile)) {
            String line;

            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(",");
                if (parts.length > 2) {
                    allCommittedFiles.addAll(Arrays.asList(parts[3].split(" ")));
                    allCommittedHashes.addAll(Arrays.asList(parts[4].split(" ")));
                }
            }
        }

        Map<String, String> dictionary = new HashMap<>();
        Map<String, String> flippedDict = new HashMap<>();
        for (int i = 0; i < Math.min(allCommittedHashes.size(), allCommittedFiles.size()); i++) {
            dictionary.put(allCommittedHashes.get(i), allCommittedFiles.get(i));
        }
        for (Map.Entry<String, String> entry : dictionary.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();
            flippedDict.put(value, key);
        }

        List<String> committedHashes = new ArrayList<>(flippedDict.values());
        List<String> committedFiles = new ArrayList<>(flippedDict.keySet());
    
        for (Path file : workingFiles) {
            String relativePath = repoPath.relativize(file).toString();
            if (ignoredFiles.contains(relativePath)) {
                continue;
            }
            
            String fileHash = calculateHash(file);
            if (stagedFiles.containsKey(relativePath)) {
                if (stagedFiles.get(relativePath).equals(fileHash)) {
                    System.out.println("Staged: " + relativePath);
                } else {
                    System.out.println("Modified (staged): " + relativePath);
                }
            } else if (committedFiles.contains(relativePath)) {
                if (committedHashes.contains(fileHash)) {
                    System.out.println("Committed: " + relativePath);
                } else {
                    System.out.println("Modified and not staged: " + relativePath);
                }
            } else {
                System.out.println("Untracked: " + relativePath);
            }
        }
    }

    private static Set<String> loadTigignore(Path tigignorePath) throws IOException {
        if (!Files.exists(tigignorePath)) {
            return Collections.emptySet();
        }

        return Files.lines(tigignorePath)
                .map(String::trim)
                .filter(line -> !line.isEmpty())
                .collect(Collectors.toSet());
    }

    private static void checkout(String commitId) throws IOException {
        Path repoPath = Paths.get(System.getProperty("user.dir"));
        Path tigDir = repoPath.resolve(".tig");
        Path commitDir = tigDir.resolve(commitId);
        Path commitsFile = tigDir.resolve(COMMIT_FILE);

        if (!Files.exists(commitDir)) {
            System.err.println("Error: Commit ID not found: " + commitId);
            return;
        }

        Files.list(commitDir).filter(Files::isRegularFile).forEach(file -> {
            try {
                List<String> committedFiles = new ArrayList<>();
                List<String> committedHashes = new ArrayList<>();

                try (BufferedReader reader = Files.newBufferedReader(commitsFile)) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        String[] parts = line.split(",");
                        if (parts.length > 0 && parts[0].equals(commitId)) {
                            committedFiles.addAll(Arrays.asList(parts[3].split(" ")));
                            committedHashes.addAll(Arrays.asList(parts[4].split(" ")));
                        }
                    }
                }
                Map<String, String> dictionary = new HashMap<>();
                for (int i = 0; i < Math.min(committedHashes.size(), committedFiles.size()); i++) {
                    dictionary.put(committedHashes.get(i), committedFiles.get(i));
                }
                String hashName = file.getFileName().toString().split("\\.")[0];
                String originalName = dictionary.get(hashName);
                Path destPath = repoPath.resolve(originalName);
                Files.copy(file, destPath, StandardCopyOption.REPLACE_EXISTING);
            } catch (IOException e) {
                e.printStackTrace();
            }
        });

        System.out.println("Checked out commit: " + commitId);
    }

    private static void log(String... args) throws IOException {
        Path currentPath = Paths.get(System.getProperty("user.dir"));
        
        if (!currentPath.endsWith(REPO_DIR)) {
            currentPath = currentPath.resolve(REPO_DIR);
        }
        
        Path tigDir = currentPath.resolve(".tig");
        Path commitsFile = tigDir.resolve(COMMIT_FILE);
    
    
        if (!Files.exists(commitsFile)) {
            System.out.println("there are no commits found.");
            return;
        }
    
        List<String> commits = Files.lines(commitsFile).collect(Collectors.toList());
    
        //commits.forEach(System.out::println);
    
        if (commits.isEmpty()) {
            System.out.println("No commits found.");
            return;
        }
    
        int n = 5; // default value
    
        if (args.length >= 1) {
            n = Integer.parseInt(args[0].substring(1)); 
        }
    
        int start = Math.max(0, commits.size() - n);
        for (int i = commits.size() - 1; i >= start; i--) {
            System.out.println();
            String[] parts = commits.get(i).split(",");
            System.out.println("Commit ID: " + parts[0]);
            System.out.println("Date: " + parts[1]);
            System.out.println("Message: " + parts[2]);
            System.out.println();
        }
    }
    
    private static void diff(String fileName) throws IOException {
        Path repoPath = Paths.get(System.getProperty("user.dir"));
        Path tigDir = repoPath.resolve(".tig");
        Path commitsFile = tigDir.resolve(COMMIT_FILE);
    
        if (!Files.exists(commitsFile)) {
            System.out.println("Error: There are no commits done. Commit a file first.");
            return;
        }
    
        List<String> commits = Files.readAllLines(commitsFile, StandardCharsets.UTF_8);
        String latestCommitId = null;
        String committedHash = null;
    
        for (int i = commits.size() - 1; i >= 0; i--) {
            String[] parts = commits.get(i).split(",");
            String commitId = parts[0];
            String[] files = parts[3].split(" ");
            String[] hashes = parts[4].split(" ");
    
            for (int j = 0; j < files.length; j++) {
                if (files[j].equals(fileName)) {
                    latestCommitId = commitId;
                    committedHash = hashes[j];
                    break;
                }
            }
    
            if (committedHash != null) {
                break;
            }
        }
    
        if (committedHash == null) {
            System.out.println("Error: The file: " + fileName + "  was not found in any commit.");
            return;
        }
    
        Path committedFilePath = tigDir.resolve(latestCommitId).resolve(committedHash + ".bck");
        Path currentFilePath = repoPath.resolve(fileName);
        List<String> committedLines;
        List<String> currentLines;
    
        // issue reading the file in utf-8, tried different formats 
        try {
            committedLines = Files.readAllLines(committedFilePath, StandardCharsets.UTF_8);
            currentLines = Files.readAllLines(currentFilePath, StandardCharsets.UTF_8);
        } catch (MalformedInputException e) {
            System.out.println("Issues found in UTF-8. Trying UTF-16.");
    
            try {
                committedLines = Files.readAllLines(committedFilePath, StandardCharsets.UTF_16);
                currentLines = Files.readAllLines(currentFilePath, StandardCharsets.UTF_16);
            } catch (MalformedInputException ex) {
                System.out.println("Issues found in UTF-16. Trying binary.");
    
                // if UTF-16 fails, read as binary
                committedLines = readAsBinaryLines(committedFilePath);
                currentLines = readAsBinaryLines(currentFilePath);
            }
        }
    
        generateUnifiedDiff(fileName, committedLines, currentLines);
    }
    
    private static void generateUnifiedDiff(String fileName, List<String> committedLines, List<String> currentLines) {
        int committedLength = committedLines.size();
        int currentLength = currentLines.size();
        int maxLength = Math.max(committedLength, currentLength);
    
        List<String> diffResult = new ArrayList<>();
        //int contextLines = 5; 
    
        for (int i = 0; i < maxLength; i++) {
            String committedLine = (i < committedLength) ? committedLines.get(i) : null;
            String currentLine = (i < currentLength) ? currentLines.get(i) : null;
    
            if (committedLine == null) {
                diffResult.add("+ Line " + (i + 1) + ": " + currentLine);
            } else if (currentLine == null) {
                diffResult.add("- Line " + (i + 1) + ": " + committedLine);
            } else if (!committedLine.equals(currentLine)) {
                diffResult.add("- Line " + (i + 1) + ": " + committedLine);
                diffResult.add("+ Line " + (i + 1) + ": " + currentLine);
            } else {
                if (i == 0 || i == maxLength - 1 || (!committedLines.get(i).equals(currentLines.get(i - 1)) && !committedLines.get(i).equals(currentLines.get(i + 1)))) {
                    diffResult.add("  Line " + (i + 1) + ": " + committedLine);
                }
            }
        }
    
        if (diffResult.isEmpty()) {
            System.out.println("There is no difference between the current file and the committed version.");
        } else {
            System.out.println("Diff between committed and current version of " + fileName + ":");
            diffResult.forEach(System.out::println);
        }
    }
    
    private static List<String> readAsBinaryLines(Path filePath) throws IOException {
        List<String> lines = new ArrayList<>();
        byte[] fileContent = Files.readAllBytes(filePath);
        StringBuilder sb = new StringBuilder();
    
        for (byte b : fileContent) {
            sb.append(String.format("%02x ", b));
        }
    
        lines.add(sb.toString().trim());
        return lines;
    }

    
    // private static void diff(String fileName) throws IOException {
    //     Path repoPath = Paths.get(System.getProperty("user.dir"));
    //     Path tigPath = repoPath.resolve(".tig");

    //     List<Path> manifestFiles = Files.walk(tigPath)
    //             .filter(file -> file.toString().endsWith(".csv"))
    //             .collect(Collectors.toList());
    //     manifestFiles.sort((file1, file2) -> file2.compareTo(file1));

    //     Path latestManifest = manifestFiles.get(0);
    //     String committedHash = null;

    //     try (BufferedReader reader = Files.newBufferedReader(latestManifest)) {
    //         String line;
    //         while ((line = reader.readLine()) != null) {
    //             String[] parts = line.split(",");
    //             if (parts[0].equals(fileName)) {
    //                 committedHash = parts[1];
    //                 break;
    //             }
    //         }
    //     }

    //     if (committedHash == null) {
    //         System.out.println("File not found in any commit.");
    //         return;
    //     }

    //     Path committedDir = tigPath.resolve(committedHash);
    //     Path committedFile = committedDir.resolve(fileName);
    //     if (!Files.exists(committedFile)) {
    //         System.out.println("Committed file not found: " + committedFile);
    //         return;
    //     }

    //     Path currentFile = repoPath.resolve(fileName);
    //     if (!Files.exists(currentFile)) {
    //         System.out.println("Current file not found: " + currentFile);
    //         return;
    //     }

    //     try (BufferedReader committedReader = Files.newBufferedReader(committedFile);
    //          BufferedReader currentReader = Files.newBufferedReader(currentFile)) {
    //         String committedLine;
    //         String currentLine;
    //         while ((committedLine = committedReader.readLine()) != null) {
    //             currentLine = currentReader.readLine();
    //             if (currentLine == null) {
    //                 System.out.println(committedLine);
    //             } else if (!committedLine.equals(currentLine)) {
    //                 System.out.println(committedLine);
    //                 System.out.println("+" + currentLine);
    //             }
    //         }

    //         while ((currentLine = currentReader.readLine()) != null) {
    //             System.out.println("+" + currentLine);
    //         }
    //     }
    // }

    // private static void log(String num, String... args) throws IOException {
    //     Path commitsFile = Paths.get(System.getProperty("user.dir"), REPO_DIR, ".tig", COMMIT_FILE);
    //     List<String> commits = Files.lines(commitsFile).collect(Collectors.toList());

    //     int start = Math.max(0, commits.size() - Integer.parseInt(num));
    //     for (int i = commits.size() - 1; i >= start; i--) {
    //         String[] parts = commits.get(i).split(",");
    //         System.out.println("Commit ID: " + parts[0]);
    //         System.out.println("Date: " + parts[1]);
    //         System.out.println("Message: " + parts[2]);
    //         System.out.println();
    //     }
    // }

    private static String calculateHash(Path file) throws IOException {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] bytes = Files.readAllBytes(file);
            byte[] hashBytes = digest.digest(bytes);
            return bytesToHex(hashBytes, HASH_LEN);
        } catch (NoSuchAlgorithmException e) {
            throw new IOException("Unable to calculate hash.", e);
        }
    }

    private static String bytesToHex(byte[] bytes, int length) {
        StringBuilder hexString = new StringBuilder();
        for (int i = 0; i < length && i < bytes.length; i++) {
            String hex = Integer.toHexString(0xff & bytes[i]);
            if (hex.length() == 1) hexString.append('0');
            hexString.append(hex);
        }
        return hexString.toString();
    }

    private static Map<String, String> loadCsvToMap(String filename) throws IOException {
        // Path path = Paths.get(REPO_DIR, filename);
        Path path = Paths.get(filename);
        if (!Files.exists(path)) return new HashMap<>();

        Map<String, String> map = new HashMap<>();
        for (String line : Files.readAllLines(path)) {
            String[] parts = line.split(",");
            map.put(parts[0], parts[1]);
        }
        return map;
    }

    private static void saveMapToCsv(String filename, Map<String, String> map) throws IOException {
        // Path path = Paths.get(REPO_DIR, filename);
        Path path = Paths.get(filename);
        try (BufferedWriter writer = Files.newBufferedWriter(path)) {
            for (Map.Entry<String, String> entry : map.entrySet()) {
                writer.write(entry.getKey() + "," + entry.getValue());
                writer.newLine();
            }
        }
    }
}

File: tig.py
import csv
import shutil
import sys
import time
import json
import uuid
from pathlib import Path
from glob import glob
from hashlib import sha256
import os
import difflib

HASH_LEN = 16



BASE_DIR = Path(__file__).parent
REPO_DIR = BASE_DIR / 'repo'
BACKUP_DIR = REPO_DIR / '.tig'
SOURCE_DIR = REPO_DIR

STAGE_FILE = BACKUP_DIR / "staged_state.json"
COMMIT_FILE = BACKUP_DIR / "commits.json"

def hash_all(root):
    """
    Create hash values for all files in the repository.
    """
    result = []
    for name in glob(str(root / "**/*.*"), recursive=True):
        full_name = Path(name)
        with open(full_name, "rb") as reader:
            data = reader.read()
            hash_code = sha256(data).hexdigest()[:HASH_LEN]
            result.append((str(full_name.relative_to(root)), hash_code))
    return result

def copy_files(source_dir, backup_dir, manifest):
    """
    Copy files to the backup (.tig) directory using their hash codes as filenames.
    """
    for (filename, hash_code) in manifest:
        source_path = Path(source_dir, filename)
        backup_path = Path(backup_dir, f"{hash_code}.bck")
        if not backup_path.exists():
            shutil.copy(source_path, backup_path)

def current_time():
    return f"{time.time()}".split(".")[0]

def write_manifest(backup_dir, timestamp, manifest):
    """
    Write a manifest file listing the files and their hashes.
    """
    backup_dir = create_dir(backup_dir)
    manifest_file = Path(backup_dir, f"{timestamp}.csv")
    with open(manifest_file, "w") as raw:
        writer = csv.writer(raw)
        writer.writerow(["filename", "hash"])
        writer.writerows(manifest)

def create_dir(dir_path):
    """
    Create a directory if it does not exist.
    """
    backup_dir = Path(dir_path)
    if not backup_dir.exists():
        backup_dir.mkdir()
    return backup_dir

def save_staged_state(staged_state):
    """
    Save the current staged state to a JSON file in the repo directory.
    """
    with open(REPO_DIR / STAGE_FILE, "w") as file:
        json.dump(staged_state, file)

def load_staged_state():
    """
    Load the current staged state from a JSON file in the repo directory.
    """
    stage_file_path = REPO_DIR / STAGE_FILE
    if stage_file_path.exists():
        with open(stage_file_path, "r") as file:
            staged = json.load(file)
        staged = [tuple(item) for item in staged]
        return staged
    else:
        return []

def add_file_to_stage(file_path):
    """
    Add a file to the staging area.
    """
    untracked_file = Path(REPO_DIR, file_path)
    assert untracked_file.exists(), f'Error: the file: {untracked_file} does not exist in this folder'

    with open(untracked_file, "rb") as reader:
        data = reader.read()
        hash_code = sha256(data).hexdigest()[:HASH_LEN]

    staged_state = load_staged_state()
    for i, (staged_file, hash_val) in enumerate(staged_state):
        if staged_file == file_path:
            if hash_val == hash_code:
                print(f'the file {untracked_file} is already in the staged state')
                return
            else:
                print(f'the file {staged_file} was updated in the staged state')
                staged_state[i] = (file_path, hash_code)
                save_staged_state(staged_state)
                return

    staged_state.append((file_path, hash_code))
    print(f'the file {file_path} was added to the staged state')

    save_staged_state(staged_state)

def commit(commit_msg):
    """
    Commit the current staged files, saving them to the .tig directory.
    """
    commit_id = str(uuid.uuid4())[:HASH_LEN]
    commit_date = time.strftime('%Y-%m-%d %H:%M:%S', time.gmtime(time.time()))
    timestamp = current_time()

    staged_state = load_staged_state()
    assert staged_state, "No files to commit. Nothing to change"

    for file_path, hash_code in staged_state:
        source_path = REPO_DIR / file_path
        dest_path = BACKUP_DIR / f"{hash_code}.bck"
        shutil.copy(source_path, dest_path)

    write_manifest(BACKUP_DIR, timestamp, staged_state)

    commits_file = REPO_DIR / COMMIT_FILE
    commits = []
    if commits_file.exists():
        with open(commits_file, "r") as file:
            commits = json.load(file)

    new_commit = {
        "id": commit_id,
        "date": commit_date,
        "message": commit_msg,
        "files": {file: hash for file, hash in staged_state},
        "timestamp": timestamp
    }
    commits.append(new_commit)

    with open(commits_file, "w") as file:
        json.dump(commits, file, indent=4)

    save_staged_state([])

    print(f"Commit '{commit_msg}' is completed with ID {commit_id}")

def log(num=5):
    """
    Display the log of previous commits.
    """
    commits_file = REPO_DIR / COMMIT_FILE
    with open(commits_file, "r") as file:
        commits = json.load(file)

        print("\nCommit History:\n")
        for i, commit in enumerate(reversed(commits)):
            if i == num:
                break
            print(f"{i + 1}. Commit ID: {commit['id']}")
            print(f"   Date: {commit['date']}")
            print(f"   Message: {commit['message']}\n")

def get_status():
    """
    Get the status of all files in the repo directory.
    """
    all_files = hash_all(SOURCE_DIR)
    staged_files = load_staged_state()
    excluded_files = {STAGE_FILE, COMMIT_FILE}
    #staged_files_dict = {file: hash for file, hash in staged_files}
    
    try:
        with open(COMMIT_FILE, 'r') as file:
            commit_data = json.load(file)
        
        commited_files = {}
        for commit in commit_data:
            for file in commit["files"]:
                commited_files[file] = commit["files"][file]
    except:
        commited_files = {}


    for file, hash in all_files:
        if file in excluded_files:
            continue
    
        if file in commited_files and hash == commited_files[file]:
            print(f"The file '{file}' is commited.")
        elif (file, hash) in staged_files:
            print(f"The file '{file}' is staged.")
        elif file in commited_files and hash != commited_files[file]:
            print(f"The file '{file}' is modified and not staged.")
        else:
            print(f"The file '{file}' is untracked.")

def checkout(commit_id):
    """
    Checkout a commit by restoring all files to their committed state.
    """
    commits_file = REPO_DIR / COMMIT_FILE
    assert commits_file.exists(), "Error: No commits found."

    with open(commits_file, "r", newline="") as file:
        commits = json.load(file)

    commit = next((c for c in commits if c["id"] == commit_id), None)
    assert commit, f"Error: Commit ID {commit_id} not found."

    for filename, hash_value in commit["files"].items():
        committed_file_path = BACKUP_DIR / f"{hash_value}.bck"
        assert committed_file_path.exists(), f"Error: Committed file {filename} not found."

        with open(committed_file_path, "r",newline="") as src, open(REPO_DIR / filename, "w",newline="") as dest:
            dest.write(src.read())

    print(f"Checked out commit {commit_id}. Files restored to the working directory.")

def diff(filename):
    """
    Show the differences between the current file and the last committed version.
    """
    manifest_files = sorted(BACKUP_DIR.glob("*.csv"), reverse=True)  
    assert manifest_files, "Error: No commits found. Commit a file first."

    latest_manifest = manifest_files[0]
    committed_hash = None

    with open(latest_manifest, "r", newline="") as manifest:
        reader = csv.DictReader(manifest)
        for row in reader:
            if row["filename"] == filename:
                committed_hash = row["hash"]
                break

    assert committed_hash, f"Error: File {filename} not found in the latest commit."

    committed_file_path = BACKUP_DIR / f"{committed_hash}.bck"
    assert committed_file_path.exists(), f"Error: Committed version of {filename} not found."

    with open(committed_file_path, "r", newline="", encoding="utf-16") as committed_file, open(REPO_DIR / filename, "r", newline="", encoding="utf-16") as current_file:
        committed_lines = committed_file.readlines()
        current_lines = current_file.readlines()

    diff_result = difflib.unified_diff(
        committed_lines,
        current_lines,
        fromfile=f"Committed: {filename}",
        tofile=f"Working: {filename}",
        lineterm=""
    )
    
    
    result = "\n".join(diff_result)
    if not result:
        result = "There is no difference in the file and the previous version."
    print(result)

if __name__ == "__main__":
    if not REPO_DIR.exists():
        print("Error: Repository folder 'repo' does not exist. Please create it first.")
        sys.exit(1)

    command = sys.argv[1]

    if command == 'init':
        create_dir(BACKUP_DIR)
        print(f"Initialized empty repository in {BACKUP_DIR}")
    elif command == 'add':
        assert len(sys.argv) > 2, "Usage: tig.py add <file_path>"
        add_file_to_stage(sys.argv[2])
    elif command == 'commit':
        assert len(sys.argv) > 2, "Usage: tig.py commit <message>"
        commit(sys.argv[2])
    elif command == 'log':
        num = 5 if len(sys.argv) == 2 else int(sys.argv[2].split('-')[-1])
        log(num)
    elif command == 'status':
        get_status()
    elif command == 'checkout':
        assert len(sys.argv) > 2, "Usage: tig.py checkout <commit_id>"
        checkout(sys.argv[2])
    elif command == 'diff':
        assert len(sys.argv) > 2, "Usage: tig.py diff <filename>"
        diff(sys.argv[2])